sbncode/BeamSpillInfoRetriever/NuMIRetriever/NuMIRetriever_module.cc

```diff
@@ -0,0 +1,254 @@
+/*
+ * NuMI Beam Spill Info Retriever module for ICARUS
+ * Based heavily on code by Z. Pavlovic written for MicroBooNE
+ * Based heavily on code by NOvA collaboration (Thanks NOvA!)
+*/
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "lardata/Utilities/AssociationUtil.h"
+#include "lardataobj/Utilities/sparse_vector.h"
+#include "larcoreobj/SimpleTypesAndConstants/RawTypes.h"
+#include "larcorealg/Geometry/Exceptions.h"
+
+#include "artdaq-core/Data/Fragment.hh"
+#include "sbndaq-artdaq-core/Overlays/ICARUS/ICARUSTriggerUDPFragment.hh"
+
+#include "lardataalg/DetectorInfo/DetectorPropertiesStandard.h"
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "sbnobj/Common/POTAccounting/NuMISpillInfo.h"
+
+#include "IFBeam_service.h"
+#include "ifbeam_c.h"
+//#include "MWRData.h"
+
+#include <memory>
+#include <optional>
+#include <vector>
+
+namespace sbn {
+  class NuMIRetriever;
+}
+
+class sbn::NuMIRetriever : public art::EDProducer {
+public:
+  explicit NuMIRetriever(fhicl::ParameterSet const &p);
+  NuMIRetriever(NuMIRetriever const&) = delete;
+  NuMIRetriever(NuMIRetriever &&) = delete;
+  NuMIRetriever& operator=(NuMIRetriever const&) = delete;
+  NuMIRetriever& operator=(NuMIRetriever&&) = delete;
+  
+  // Required functions.
+  void produce(art::Event& e) override;
+  void beginSubRun(art::SubRun& sr);
+  void endSubRun(art::SubRun& sr);
+  
+private:
+  // input labels
+  std::vector< sbn::NuMISpillInfo > fOutbeamInfos;
+  double fTimePad;
+  double fBFPEpsilion;
+  std::string fURL;
+  //MWRData mwrdata;
+  std::string raw_data_label_;
+  std::string fDeviceUsedForTiming;
+  int TotalBeamSpills;
+  art::ServiceHandle<ifbeam_ns::IFBeam> ifbeam_handle;
+  std::unique_ptr<ifbeam_ns::BeamFolder> bfp;
+};
+
+sbn::NuMIRetriever::NuMIRetriever(fhicl::ParameterSet const& p)
+  : EDProducer{p},
+  fTimePad(p.get<double>("TimePadding", 0.5)), //epsilon in seconds, buffer time to look for spills
+  fBFPEpsilion(p.get<double>("BFPEpsilon", 0.02)), // 20 ms, tuned for BNB, check for NuMI here might need to be larger
```

Comment:
Is this too small? Have you checked that you aren't getting very very few matches? I think Andrea had some runs where they triggered only on NuMI and or triggered on every NuMI gate

Reply:
Yeah I'm gonna tune the epsilon a bit -- I'll fix the value in a future PR.

Reply:
OK, sounds good.